File: android/src/main/kotlin/com/lkrjangid/account_manager/AccountManagerPlugin.kt
package com.lkrjangid.account_manager

import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel

/** AccountManagerPlugin — Pigeon-based Flutter plugin entry point. */
class AccountManagerPlugin : FlutterPlugin {

    private var hostApiImpl: AccountManagerHostApiImpl? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        setup(binding.binaryMessenger, binding.applicationContext)
    }

    private fun setup(messenger: BinaryMessenger, context: Context) {
        val impl = AccountManagerHostApiImpl(context)
        hostApiImpl = impl
        AccountManagerHostApi.setUp(messenger, impl)

        // No-op config channel — keychainAccessGroup is iOS-only.
        // Registering the handler prevents MissingPluginException on the Dart side.
        MethodChannel(messenger, "flutter_account_manager/config")
            .setMethodCallHandler { _, result -> result.success(null) }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        AccountManagerHostApi.setUp(binding.binaryMessenger, null)
        hostApiImpl?.dispose()
        hostApiImpl = null
    }
}

File: ios/Classes/AccountManagerPlugin.swift
import Flutter
import UIKit

/// Flutter plugin entry point for the account_manager plugin.
public class AccountManagerPlugin: NSObject, FlutterPlugin {

    /// Retained for the lifetime of the plugin so the configure channel handler
    /// can update the active KeychainManager access group.
    private static var sharedImpl: AccountManagerHostApiImpl?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let messenger = registrar.messenger()
        let impl = AccountManagerHostApiImpl()
        sharedImpl = impl
        AccountManagerHostApiSetup.setUp(binaryMessenger: messenger, api: impl)

        // Secondary MethodChannel that lets the Dart layer configure the
        // Keychain Access Group before the first credential operation.
        let configChannel = FlutterMethodChannel(
            name: "flutter_account_manager/config",
            binaryMessenger: messenger
        )
        configChannel.setMethodCallHandler { call, result in
            guard call.method == "configure" else {
                result(FlutterMethodNotImplemented)
                return
            }
            let args = call.arguments as? [String: Any?]
            let group = args?["keychainAccessGroup"] as? String
            AccountManagerPlugin.sharedImpl?.configure(accessGroup: group)
            result(nil)
        }
    }
}
